src/main/java/com/google/api/gax/grpc/AbstractPage.java

```diff
@@ -0,0 +1,171 @@
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.gax.core.ApiFunction;
+import com.google.api.gax.core.ApiFuture;
+import com.google.api.gax.core.ApiFutures;
+import com.google.api.gax.core.AsyncPage;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.Iterables;
+import java.util.Iterator;
+
+public abstract class AbstractPage<
+        RequestT,
+        ResponseT,
+        ResourceT,
+        PageT extends AbstractPage<RequestT, ResponseT, ResourceT, PageT>>
+    implements AsyncPage<ResourceT> {
+
+  private final PageContext<RequestT, ResponseT, ResourceT> context;
+  private final ResponseT response;
+
+  protected AbstractPage(PageContext<RequestT, ResponseT, ResourceT> context, ResponseT response) {
+    this.context = Preconditions.checkNotNull(context);
+    this.response = Preconditions.checkNotNull(response);
+  }
+
+  protected abstract PageT createPage(
+      PageContext<RequestT, ResponseT, ResourceT> context, ResponseT response);
+
+  @Override
+  public boolean hasNextPage() {
+    return !getNextPageToken().equals(context.pageDescriptor().emptyToken());
+  }
+
+  @Override
+  public String getNextPageToken() {
+    return context.pageDescriptor().extractNextToken(response);
+  }
+
+  @Override
+  public PageT getNextPage() {
+    if (hasNextPage()) {
+      RequestT nextRequest =
+          context.pageDescriptor().injectToken(context.request(), getNextPageToken());
+      ResponseT response =
+          ApiExceptions.callAndTranslateApiException(
+              context.callable().futureCall(context.request(), context.callContext()));
+      return createPage(context.withRequest(nextRequest), response);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public ApiFuture<PageT> getNextPageAsync() {
+    if (hasNextPage()) {
+      RequestT nextRequest =
+          context.pageDescriptor().injectToken(context.request(), getNextPageToken());
+      final PageContext<RequestT, ResponseT, ResourceT> newContext =
+          context.withRequest(nextRequest);
```

Comment:
Maybe move `nextRequest` and `newContext` into the `apply` function below (i.e. construct them right before they're needed)